### ISSUE TITLE
CWG2633 [expr.const] p7 The comment seems to be wrong 

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#124]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

Consider the example in [expr.const] p7
````cpp
extern Swim dc;
constexpr auto& sandeno   = typeid(dc);         // OK, can only be typeid(Swim)
````
According to [dcl.constexpr] p7, the full-expression of the initialization shall be a constant expression. `dc` is a glvalue that refers to an object, and according to [expr.const] p7 that says
> During the evaluation of an expression E as a core constant expression, **all id-expressions** and uses of *this that refer to **an object** or reference **whose lifetime did not begin with the evaluation of E** are treated as referring to a specific instance of that object or reference whose lifetime and that of all subobjects (including all union members) includes the entire constant evaluation. **For such an object that is not usable in constant expressions**, the dynamic type of the object is **constexpr-unknown**.

`dc` is an **id-expression** that refers to an object and the object is not usable in constant expression, and the lifetime of the object whatever did not begin with the evaluation of the full-expression of the initialization(i.e `E`). Moreover, [expr.const] p5.26 says
> a dynamic_­cast ([expr.dynamic.cast]) or typeid ([expr.typeid]) expression **on a glvalue that refers to an object whose dynamic type is constexpr-unknown** or that would throw an exception;

the operand `dc` of the expression `typeid` is the object whose dynamic type is constexpr-unknown, hence  `typeid(dc)` shouldn't be a core constant expression as per the above rules. The comment seems to be wrong, or to say there is not a sufficient rule to say that. 

### Suggested resolution
I suppose the intent of [expr.const] p5.26 should be
> a dynamic_­cast ([expr.dynamic.cast]) or typeid ([expr.typeid]) expression on a glvalue that is not an id-expression that denotes an object([expr.prim.id.unqual]p3, [expr.prim.id.qual] p5), where the glvalue refers to an object whose dynamic type is constexpr-unknown, or the expression would throw an exception;

In other words, the operand(glvalue) we are talking about is either a reference or the result of an indirection of a pointer.

**Comment by jensmaurer** [comment]
CWG2633
